bpf: daemon: Replace run_probes.sh with Golang code

vadorovsky opened the pull request:
Implement all BPF probes as a part of Cilium code instead of
using Bash script.

Related-to: #1600
Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6290)
<!-- Reviewable:end -->

aanm commented on daemon/probes.go:
```diff
@@ -0,0 +1,160 @@
+			return err
+		}
+		if strings.HasSuffix(info.Name(), ".t") {
+			copyFile(path, outFilename)
```
You are ignoring the error returned by this function.

aanm reviewed (commented):
Looks good so far

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+	// File descriptors for feature, info and warning file.
+	featureFile, err := os.OpenFile(featureFilePath, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		return err
```
Please extend the error with the context about which path cannot be opened (here and for subsequent error returns).

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
```
I think that this code could live under `pkg/datapath/probes` or similar, there shouldn't be too much dependency on other bits in the `daemon/` directory right?

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+				return err
+			}
+			out, err := exec.Command(
+				"clang",
```
Since you're just trying to translate the existing code into Golang, I can see why this is written like this, but I wonder if there's a specific reason why we need to compile these at runtime, rather than just shipping precompiled binaries and only invoking them here? (This could be followed up in a separate follow-up PR to simplify review)

joestringer replied:
I note that the bpf syscall number is architecture dependent, but presumably that also affects the tc binary that we ship so presumably the container image that we ship is only appropriate for x86_64 anyway.

vadorovsky replied:
Good idea.

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+				path.Join(probesDir, "raw_main.c"),
+				"-o",
+				outBin,
+			).CombinedOutput()
```
We've hit problems before where a step like this takes an inordinate amount of time (for example, when someone sets very low CPU constraints on a cilium deployment). I think that at least for the `CombinedOutput()` call here we could consider replacing with the library wrapper in `pkg/command/exec` which uses a context/timeout to bound the execution time. This is not such a big deal for probes like this but might be helpful to know.

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+		return err
+	}
+
+	// Create tempotary dir for probes.
```
This comment is misspelled, and states exactly what the line underneath says in Go code. Not sure it's that helpful?

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+	}
+	warningFile, err := os.OpenFile(warningFilePath, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		return err
```
In this error condition, the `infoFile` is not closed.

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+	// Create tempotary dir for probes.
+	outDir, err := ioutil.TempDir("", "cilium-probes-out")
+	if err != nil {
+		return fmt.Errorf("could not create temporary directory for probes: %s", err)
```
In this condition, neither the info or warning files are closed.

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+	// to info and warning file.
+	infoFileBytes, warningFileBytes, err := writeFeatures(featureFile, infoFile, warningFile, probesDir, libIncludeDir, outDir)
+	if err != nil {
+		return err
```
Again, leaking files

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+
+// runProbes runs probes for BPF features. It generates a header file for BPF
+// programs and log files.
+func runProbes(bpfDir, stateDir string) error {
```
Most of the code in this function seems to be the same operation for two or three files (rm, open, close, sync), but copied and pasted. I wonder if it would be a little easier to ensure that they are exactly the same by, for example, refactoring some of the common sets of operations.

vadorovsky replied:
I tried my best. I hope that the `fileWithCleanup` structure is abstract enough.

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+	if err != nil {
+		return err
+	}
+	defer featureFile.Close()
```
`Close()` can fail, and I see at least some documents online suggesting that it's poor style to defer file closes when you write to files, particularly in this form because the error is lost.

joestringer commented on daemon/probes.go:
```diff
@@ -0,0 +1,290 @@
+	return
+}
+
+// writeFeatures wrtites definitions of features into the header file which is
```
```suggestion
// writeFeatures writes definitions of features into the header file which is
```

joestringer reviewed (changes requested):
I realise this is RFC, but overall it seems to take a useful approach of attempting to directly translate the shell logic into BPF, which eases review. It seemed pretty similar to what's there in the shell script so good enough to do a proper review on.

Please take a close look at every error return condition, and consider the state of eg the filesystem in each of these cases (in particular whether files are synced/closed), and also what the error message will look like in the log (for example, whether it has the relevant context to point out where in these functions the error occurs).

With this PR, the file `test/bpf/probes-test.sh` could also be removed. It seems like the privileged unit test here has effectively the same test logic.

You might also consider adding a unit test for `probeKernelConfig()`.

Wherever this logic ends up, it would also be good to add it to CODEOWNERS with the bpf group as the reviewers.

nebril commented on pkg/datapath/probes/probes.go:
```diff
@@ -0,0 +1,347 @@
+// Sync commits the current contents of the file to stable storage. Typically,
+// this means flushing the file system's in-memory copy of recently written
+// data to disk.
+func (f *fileWithCleanup) Sync() error {
```
Can you just return f.File.Sync here? it doesn't seem like this implementation is doing anything else.

vadorovsky replied:
I just wanted to extend the error message. I think it's a bit cleaner than writing

```
if err != nil {
    return fmt.Errorf("could not sync file %s: %s", f.path, err)
}
```

multiple times. With that override, I can just do `if err != nil { return err }` later.

nebril commented on pkg/datapath/probes/probes.go:
```diff
@@ -0,0 +1,347 @@
+	if err := f.File.Close(); err != nil {
+		return fmt.Errorf("could not close file %s: %s", f.path, err)
+	}
+	if f.n > 0 {
```
Shouldn't this be `if f.n == 0`? It seems like right now you are deleting the file if there were non-zero `n` of bytes written into it, which seems the other way around expected behaviour.

vadorovsky replied:
Yes, you're right.

raybejjani commented on pkg/datapath/probes/probes.go:
```diff
@@ -0,0 +1,355 @@
+
+// copyFile copies the file from source to destination. Before calling this
+// function, there should be no file under destination path.
+func copyFile(src, dest string) error {
```
This might be simpler if we use https://golang.org/pkg/io/ioutil/#WriteFile and ReadFile. WriteFile will do the create for us.
Was the intention to fail if the file already existed?

vadorovsky replied:
`io.Copy` copies fd contents in chunks. https://golang.org/src/io/io.go?s=12784:12844#L353 Using `WriteFile` and `ReadFile` would load the whole file content into memory at once. So IMO it's more convenient to use `io.Copy`. The amount of LOC will be similar regardless of which function I will use. :)

>Was the intention to fail if the file already existed?

It was my intention, but I should probably change it. It's probably better to remove dest file if it exists.

joestringer reviewed (commented):
Though this is now passing in Travis, it still fails in Jenkins:

```
17:39:23 probes_test.go:92:
17:39:23     c.Assert(err, IsNil)
17:39:23 ... value *errors.errorString = &errors.errorString{s:"missing kernel configuration"} ("missing kernel configuration")
17:39:23 
```

joestringer commented on pkg/datapath/probes/probes.go:
```diff
@@ -0,0 +1,355 @@
+	}
+
+	if errs != nil {
+		return errs
```
I think the multierror thing here is just to allow any failures during the `Close()` operations to be logged, but it kinda complicates the return because now we need this multierror thing. However, the caller is unlikely to be able to do anything specific about a failure to `Close()`. In fact, all it does is log the message and ignore it. Wouldn't it be simpler to just log any `Close()` failures here, and propagate back the original error?

vadorovsky replied:
Done, I dropped multierror in favor of logging all errors and returning the last one.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -328,9 +298,8 @@ func checkMinRequirements() {
-	if err := exec.Command(probeScript, option.Config.BpfDir, option.Config.StateDir).Run(); err != nil {
-		log.WithError(err).Fatal("BPF Verifier: NOT OK. Unable to run checker for bpf_features")
+	if err := probes.RunProbes(option.Config.BpfDir, option.Config.StateDir); err != nil {
+		log.WithError(err).Warn("BPF Verifier: NOT OK")
```
This is different from what we currently do. Currently we `Fatal()` at this layer,  because for instance if the header file is only half-written then it will likely mess up all compilations from this point forward.

I suspect what you're trying to achieve here is to ignore the failure to fetch kernel config. If that's the case, then I think we should handle/ignore this further down in the kernel probe config, like we do in the current script.

vadorovsky replied:
OK, now I understood. Thanks for explanation! I fixed my code.

raybejjani reviewed (commented):
oops, I was meant to post this weeks ago. sorry if it doesn't make sense anymore

vadorovsky committed:
bpf: daemon: Replace run_probes.sh with Golang code

Implement all BPF probes as a part of Cilium code instead of
using Bash script.

Related-to: #1600
Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
files: Add package for file utilities

This change provides several file utilities:

- the FileWithCleanup structure which is an extension for os.File which
  removes the file if no data were written after close
- OpenOrCreate function which created or opens file in read and write
  mode
- Copy function which copies file from given source path to destination
  path; if there is any file under destination path, it gets removes

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>
